### ISSUE TITLE
Version 3.0.0

### ISSUE THREAD

**Issue by MarshallOfSound** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#171]
`3.0.0` Milestone
------------------------

## Overview
I know I only just hit the `2.0.0` milestone but I have a new target now, lets aim for **3** :+1: 

Basically the `3.0.0` milestone is going to be one massive breaking change.  Everything is being rewritten into the electron framework.  This brings a couple of benefits and a couple of caveats and I'll outline them here

**Pros**
* We can finally kick flash to the curb :smile: 
* It will be much easier for others to contribute, learning JS is a lot easier than learning C#
* Basically all features can be mirrored across but with better support for external libraries (NPM is glorious) on the electron side of the fence
* Everything I write can be built for Windows, OSX and Linux **ONE PLAYER TO RULE THEM ALL*

**Cons**
* Yeah, I added this section for the lolz, there aren't any cons :+1: 

## Upgrade Process
Moving from C# to electron in a smooth way will be no easy feat, my current plan is to have an upgrade process like so

1. Push final release of `2.X.Y`
2. Push out an update notification
3. Because electron apps install to APPDATA (hooray for portable installs) we somehow need to trigger an uninstall of the original GPMDP
4. We do this by requesting UAC and running the uninstaller in silent mode
5. Once installed updates will be handled internally by Squirrel

## What needs to be done
At some point over the next few days I'll be moving my work from `Nucleus Player` into a `3.0.0` branch in this repository.  Then Nucleus Player will cease to exist.  Once that move is complete I need to finish off a whole lot of features before even considering pushing out the upgrade

### Minimum Requirements :fallen_leaf: 
* [x] Ability to listen to music
* [x] Global hotkeys
  * [x] Standard media keys work
  * [x] Custom media keys? #37 
* [x] Last FM support
  * [x] Log in through OAuth (more secure than user + pass)
  *  ~~Show currently scrobbled status on screen~~ *We should not replicate last.fm in app*
  *  ~~View Last FM feed in App~~ *We should not replicate last.fm in app*
* [x] Dark Theme support
  * [x] Highlight color customization
* [x] Mini Player support
  * [x] By transforming the main window
  * ~~As a pop up from the system tray~~ *Move to `3.1.0` release*
* [x] Custom Interface
  * [x] Remove unneeded google UI components
  *  ~~Make drag and drop upload work~~ *Requires a chrome extension*
* [x] Desktop notifications
  * [x] Mac *(Wrap to force silent)*
  * [x] Windows 7 *(Implement manually)*
  * [x] Windows 8 *(Wrap to force silent)*
  * [x] Windows 10 *(Wrap to force silent)*
  * [x] Ubuntu
* [x] Taskbar Media Controls on Windows
* [x] Settings
  * [x] Store window location
  * [x] Store window size
  * [x] Store highlight color
  * [x] Persist browser cache / cookies
  * [x] Mini player controls on hover / always
* [x] Upgrade path from GPMDP - *win32 only*
  * [x] Check if GPMDP is installed
  * [x] Attempt to trigger the uninstall (need to trigger `UAC`)
  * [x] Minimum delete files
* [ ] Extra UI
  * [ ] Forward / Back buttons
  * [x] Introduction to the app *(A welcome screen / whats new)*
* [x] Auto Updater - Squirrel
  * [x] Windows
  * [x] Mac OSX
  * [x] Multiple streams (`stable` and `beta`)
  * [x] Add a sanity HTTP check to ensure the update server is operating
* [x] Internal Errors
  * [x] Capture internal errors and display in a desktop notification
  * [x] Notification when clicked should open the GitHub issue page (and prefill a bunch of info)
* [x] Testing
  * [ ] Setup mocha unit testing suite
    * [x] Integrate into CI
  * [ ] Setup mocha electron suite
    * [x] Integrate into CI
* [x] Publish OSX build artifacts to server
* [x] Implement the [Playback API](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/blob/master/docs/PlaybackAPI.md)
    * [x] Add playback time
* [x] Add menu for mac
    * [x] Playback (Play / Pause, Previous, Next)
    * [x] Preferences
    * [x] Help (Issues, Contribute, Donate)

### Nice to Have :beginner: 
* [x] Audio output selector support #47 
* [ ] Visualizer
    * [ ] Custom Visualizer
    * [ ] Full screen visualizer
* [x] Equalizer
    * [x] Turn off and on
    * [x] Basic equalizer sliders
    * [ ] Equalizer presets (take the ones from the standard android eq)
* [x] Close to task bar
    * [x] When the app is "closed" only minimize it to task bar
    * [x] When this happens explain what is going on
    * [x] Give the user the option to "never tell me again"
    * [x] Allow the user to disable this action in the settings page

### Testing Guidelines
* All internal Electron `main` process code should be unit tested.  We use mocha
* All internal Electron `renderer` process code should be unit tested if possible

*If we can figure it out, UI based testing using Electron as a WebDriver target would be great.*

**NOTE:** This list will be updated and possibly extended on regulary

**Comment by Devo7v** [comment]
I noticed you have created an item under Settings for "Persistent browser cache/cookies".  Would it be possible to make this optional?  Once you release this I plan on making a portable version that I will run off a flash drive and I don't need/want my cache to continually grow taking up more and more space on my flash drive.  With 2.1.1, the cache is currently at 400MB and growing, it there was a way to flush the cache or disable it across sessions it would be very helpful.

I think it is useful to save the settings and cookie/login information across sessions, but I think saving the cache across sessions should be optional.

I'm looking forward to see what you've done with the new release.

**Comment by MarshallOfSound** [comment]
@Devo7v With this completely rewritten app I believe the need for a "portable" version will be nullified.  Electron installs go to your `AppData\Local` directory so they are inherently "portable".  No admin permissions will be needed ever to run this app

**Comment by Devo7v** [comment]
@MarshallOfSound I understand that the program will be "portable" and I want to thank you for doing that.  My concern is that the cache may become huge taking up a lot of space on the flash drive.  What I'm asking is if there is (or will be) a way to either disable or flush the cache after every use.

**Comment by Devo7v** [comment]
@MarshallOfSound Whether or not I should have at this point, I played around with the new 3.0 version and you've definitely improved upon the current release.  I think we have some disconnect between our definitions of "portable".  It seems that when you refer to portable you mean a program that doesn't require admin rights; you have absolutely accomplished this.  When I'm referring to portable I mean something I can run off of a flash drive, doesn't leave any files on the host computer, and I can take with me to any computer.

The way 3.0 is currently configured my only option is to copy all 4 directories (Local\GPMDP_3, Local\SquirrelTemp, Roaming\Google Play Music Desktop Player, and Roaming\GPMDP_STORE)  from the AppData folder to my flash drive and then take it to another computer and copy all 4 directories to the AppData folder.  Depending on how big the Roaming\Google Play Music Desktop Player gets this could take quite a bit of time to copy things back and forth.

My suggestion would be to have all 4 of these directories located within 1 folder and have everything work off of relative paths so it doesn't matter where that folder is located.  Hopefully this makes sense.  Ideally what I'm trying to do is run this entirely off of a flash drive without having to move things back and forth between the flash drive and the computer.

**Comment by MarshallOfSound** [comment]
Because I am using the `Squirrel.Windows` installer and workflows I cannot choose why my app ends up.  Squirrel will put it in AppData because that is where Squirrel apps go, the only folder I have control of is `Roaming\GPMDP_STORE` and I am planning on making that folder obselete soon anyway.

TLDR: I can't put all those folders in the same place
TLDR2: Whats the point in running it off a flash drive..........

**Comment by Devo7v** [comment]
At work we use common computers, similar to what you would have set up in a computer lab at a school.  When you log on to a computer you can do most things you would do on a normal computer, but when you log off anything you did or saved on that computer gets wiped out.  So the next time I log in I would have to re-download GPMDP and set everything up again.

To get around this I've been using PortableApps that run and save everything to a flash drive.  Even though this is a specific use case, not everybody owns a computer so whether they're working on school computers, library computers, or a cyber cafe, they can take all their programs and files with them.

**Comment by lcsnwmn** [comment]
I have a (theoretically small?) feature request. It's probably too late to work into 3.0.0, but I would love the option to have the display order for playlists in the menu to be re-arrangeable (probably via drag and drop). I know GPM doesn't do this, so I could see you not wanting to differ in that respect, but having switched from Spotify to GPM, not having the playlist I just made at the top of the list is infuriating to me. I know it appears in the Recent Activity list, but that's inconsistent and doesn't appear in the main navigation menu. 

Alternately it might be cool if there was a "recent playlists" sublist at the top of playlist list, or even just a "recent Activity" link in menu (would save 1-2 clicks and some scrolling).

Thanks for considering.

**Comment by MarshallOfSound** [comment]
This won't hit `3.0.0` but I'll look into it for a later release.  I try to keep mucking around with GPM itself to a minimum.  It breaks too regularly.

You could just name your playlist "_YOUR NAME" that would force it to the top

**Comment by MarshallOfSound** [comment]
Well this was done lol
